dan_layer/storage_sqlite/src/sqlite_shard_store_factory.rs

```diff
@@ -970,6 +1002,54 @@ impl ShardStoreWriteTransaction<PublicKey, TariDanPayload> for SqliteShardStoreW
         Ok(())
     }
 
+    fn save_current_state(
+        &mut self,
+        payload: PayloadId,
+        shard_id: ShardId,
+        leader_round: u32,
+        leader: PublicKey,
+    ) -> Result<(), StorageError> {
+        use crate::schema::current_states;
+
+        let payload_id = payload.as_bytes().to_vec();
+
+        let new_row = NewCurrentState {
+            payload_id: payload_id.clone(),
+            shard_id: shard_id.as_bytes().to_vec(),
+            leader_round: i64::from(leader_round),
+            leader: leader.as_bytes().to_vec(),
+        };
+
+        match diesel::insert_into(current_states::table)
+            .values(&new_row)
+            .execute(self.connection())
+        {
+            Ok(_) => {},
+            Err(err) => match err {
+                Error::DatabaseError(DatabaseErrorKind::UniqueViolation, _) => {
+                    match diesel::update(current_states::table)
```

Comment:
I think you can do something more idiomatic, as follows:

```rs
diesel::insert_into(current_states::table)
    .values(&new_row)
    .on_conflict(id)
    .do_update()
    .set(current_states::leader_round.eq(i64::from(leader_round)))
    .filter(current_states::payload_id.eq(payload_id))
    .execute(&conn);
 ```

Reply:
If we ever update diesel to 2.x then I can change this. The on_conflict is supported only for PG, not sqlite.